graql/reasoner/query/DisjunctiveQuery.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    public Set<ResolvableQuery> getClauses() {
+        return new HashSet<>(clauses);
+    }
+
+    public Set<Variable> getBindingVars() {
+        return new HashSet<>(bindingVars);
+    }
+
+    // Only needed for a disjunction to be the body of a rule
+    @Override
+    public Set<String> validateOntologically(Label ruleLabel) {
+        throw new UnsupportedOperationException("A disjunctions in rule body is not yet supported");
+    }
+
+    public HashMap<Variable, Concept> filterBindingVars(Map<Variable, Concept> map) {
+
+        HashMap<Variable, Concept> bindingVarsMap = new HashMap<>();
+
+        bindingVars.forEach(b -> {
+            if (map.get(b) != null) {
+                bindingVarsMap.put(b, map.get(b));
+            }
+        });
+
+        return bindingVarsMap;
+    }
+
+    @Override
+    public ResolvableQuery withSubstitution(ConceptMap sub) {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(q -> q.withSubstitution(sub)).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    public Stream<ConceptMap> traverse(){
+        Stream<Stream<ConceptMap>> answerStreams = clauses.stream().map(clause ->
+                clause.traverse().map(ans -> {
+                    ConceptMap clauseAns = new ConceptMap(ans.map(), new LookupExplanation(), clause.getPattern(ans.map()));
+                    HashMap<Variable, Concept> bindingVarsSub = filterBindingVars(ans.map());
+                    return new ConceptMap(bindingVarsSub, new DisjunctiveExplanation(clauseAns), getPattern(bindingVarsSub));
+                }));
+        LazyMergingStream<ConceptMap> mergedStreams = new LazyMergingStream<>(answerStreams);
+        return mergedStreams.flatStream();
+    }
+
+    @Override
+    ResolvableQuery inferTypes() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(ResolvableQuery::inferTypes).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery constantValuePredicateQuery() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(ResolvableQuery::constantValuePredicateQuery).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery copy() {
+        return new DisjunctiveQuery(
+                getClauses(),
+                getBindingVars(),
+                traversalExecutor,
+                context());
+    }
+
+    @Override
+    public boolean isAtomic() {
+        // TODO unclear of the meaning of atomicity in this case
+        return getClauses().stream().allMatch(ResolvableQuery::isAtomic);
+    }
+
+    @Override
+    boolean isEquivalent(ResolvableQuery q) {
```

Comment:
this is an another interesting function. I think it only carries the expected meaning for atomic queries. Might be worth to throw here to highlight unexpected execution paths.